plugins/simpleadminhacks-paper/src/main/java/com/programmerdan/minecraft/simpleadminhacks/hacks/basic/PortalModifyHack.java

```diff
@@ -40,18 +48,27 @@ public void onPlayerEnterPortal(PlayerPortalEvent event) {
 		Location to;
 		switch (getTargetWorld(player).getEnvironment()) {
 			case NETHER:
-				to = new Location(getTargetWorld(player), event.getFrom().getX(), 125, event.getFrom().getZ());
+				to = new Location(getTargetWorld(player), player.getLocation().getX(), 125, player.getLocation().getZ());
 				break;
 			case NORMAL:
 				to = getTargetWorld(player).getHighestBlockAt(player.getLocation(), HeightMap.WORLD_SURFACE).getLocation().toCenterLocation().add(0,1,0);
 				break;
 			default:
 				return;
 		}
-		event.setTo(to);
-		if (to.getWorld().getName().equals(targetWorld)) {
-			spawnExit(to);
+		event.setCancelled(true);
+		teleports.put(player, to);
+	}
+
+	@EventHandler
+	public void onTick(ServerTickEndEvent event) {
+		for (Map.Entry<Player, Location> entry : teleports.entrySet()) {
+				entry.getKey().teleport(entry.getValue(), PlayerTeleportEvent.TeleportCause.END_PORTAL);
+				if (entry.getValue().getWorld().getName().equals(targetWorld)) {
+					spawnExit(entry.getValue());
+				}
 		}
+		teleports.clear();
 	}
```

Comment:
Is it ever possible that this event can be fired simultaneously, such as during low TPS? If so I could imagine it causing some problems, and potentially more players falling through portals than before.

Reply:
Simultaneously with what? It runs at the end of a tick.